kernel/rtlil.cc

```diff
@@ -3982,23 +4012,47 @@ RTLIL::SigSpec::SigSpec(bool bit)
 {
 	cover("kernel.rtlil.sigspec.init.bool");
 
+	packed_ = false;
+	(void)new (&this->bits_) std::vector<SigBit>();
 	width_ = 0;
 	hash_ = 0;
 	append(SigBit(bit));
 	check();
 }
 
+void RTLIL::SigSpec::switch_to_packed() const
+{
+	// TODO change to debug asserts
+	log_assert(!this->packed_);
+	log_assert(bits_.size() == 0);
+	RTLIL::SigSpec *that = (RTLIL::SigSpec*)this;
+	that->bits_.~vector();
+	that->packed_ = true;
+	(void)new (&that->chunks_) std::vector<SigChunk>();
+}
+
+void RTLIL::SigSpec::switch_to_unpacked() const
+{
+	// TODO change to debug asserts
+	log_assert(this->packed_);
+	log_assert(chunks_.size() == 0);
+	RTLIL::SigSpec *that = (RTLIL::SigSpec*)this;
+	that->chunks_.~vector();
+	that->packed_ = false;
+	(void)new (&that->bits_) std::vector<SigBit>();
+}
+
 void RTLIL::SigSpec::pack() const
 {
 	RTLIL::SigSpec *that = (RTLIL::SigSpec*)this;
 
-	if (that->bits_.empty())
+	if (packed_ || that->bits_.empty())
```

Comment:
Shouldn't we be removing the `bits_.empty()` check? We need to guarantee accesses to `chunks_` are valid once `pack()` returns, so we need to do the switch in full no matter if `bits_` are empty.